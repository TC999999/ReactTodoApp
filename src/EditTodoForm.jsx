import { useState } from "react";

const EditTodoForm = ({ todo, editTodo, showEditForm }) => {
  const initialState = {
    "edit-todo": todo,
  };

  const [formData, setFormData] = useState(initialState);
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    editTodo({ ...formData });
    setFormData(initialState);
  };

  return (
    <div id="edit-todo-div">
      <form id="edit-todo-form" onSubmit={handleSubmit}>
        <label htmlFor="edit-todo">Edit Task - {todo}</label>
        <input
          id="edit-todo"
          type="text"
          name="edit-todo"
          placeholder="Enter Task Here"
          value={formData["edit-todo"]}
          onChange={handleChange}
        />

        <button>Edit Todo</button>
      </form>
      <button onClick={showEditForm}>Cancel</button>
    </div>
  );
};

export default EditTodoForm;
