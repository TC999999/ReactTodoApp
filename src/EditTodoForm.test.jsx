import { render, fireEvent } from "@testing-library/react";
import EditTodoForm from "./EditTodoForm.jsx";
import { vi } from "vitest";

it("renders without crashing", function () {
  render(<EditTodoForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<EditTodoForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("should have value and label that includes the specified task at the start", function () {
  const { getByLabelText } = render(<EditTodoForm todo="test-task" />);
  const input = getByLabelText("Edit Task - test-task");
  expect(input).toBeInTheDocument();
  expect(input).toHaveValue("test-task");
});

it("should have a changeable value", function () {
  const { getByLabelText } = render(<EditTodoForm todo="test-task" />);
  const input = getByLabelText("Edit Task - test-task");
  fireEvent.change(input, { target: { value: "new-test-task" } });
  expect(input).toHaveValue("new-test-task");
});

it("should call a function to cancel editing when the cancel button is clicked", function () {
  const showEditFormMock = vi.fn();
  const { queryByText } = render(
    <EditTodoForm todo="test-task" showEditForm={showEditFormMock} />
  );
  expect(showEditFormMock).not.toHaveBeenCalled();

  const cancel = queryByText("Cancel");
  fireEvent.click(cancel);
  expect(showEditFormMock).toHaveBeenCalled();
});

it("should empty itself out when the edit button is clicked", function () {
  const editTodoMock = vi.fn();
  const { getByLabelText, queryByText } = render(
    <EditTodoForm todo="test-task" editTodo={editTodoMock} />
  );
  expect(editTodoMock).not.toHaveBeenCalled();

  const input = getByLabelText("Edit Task - test-task");

  const btn = queryByText("Edit Todo");
  fireEvent.submit(btn);
  expect(editTodoMock).toHaveBeenCalled();
  expect(input).toBeEmptyDOMElement();
});
