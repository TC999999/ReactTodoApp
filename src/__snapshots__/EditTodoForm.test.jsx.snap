// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="edit-todo-div"
  >
    <form
      id="edit-todo-form"
    >
      <label
        for="edit-todo"
      >
        Edit Task - 
      </label>
      <input
        id="edit-todo"
        name="edit-todo"
        placeholder="Enter Task Here"
        type="text"
        value=""
      />
      <button>
        Edit Todo
      </button>
    </form>
    <button>
      Cancel
    </button>
  </div>
</DocumentFragment>
`;
